File: api/models/scrapbook.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID


class CollectibleCatalogEntry(BaseModel):
    catalog_id: int
    catalog_number: int
    plant_species_id: int
    species_name: Optional[str] = None
    scientific_name: str
    image_url: Optional[str] = None
    rarity_tier: str
    featured_order: Optional[int] = None
    is_discovered: bool
    discovered_at: Optional[datetime] = None
    discovery_id: Optional[int] = None
    user_notes: Optional[str] = None
    is_favorite: bool = False


class DiscoveryDetails(BaseModel):
    discovery_id: int
    catalog_number: int
    plant_species_id: int
    species_name: Optional[str] = None
    scientific_name: str
    family: Optional[str] = None
    rarity_tier: str
    discovered_at: datetime
    user_notes: Optional[str] = None
    is_favorite: bool = False
    plant_article: Dict[str, Any]


class DiscoveryResponse(BaseModel):
    success: bool
    message: str
    discovery_id: Optional[int] = None
    catalog_entry_id: Optional[int] = None
    catalog_number: Optional[int] = None
    species_name: Optional[str] = None
    plant_species_id: Optional[int] = None
    already_discovered: bool = False


class CollectionStats(BaseModel):
    total_collectibles: int
    total_discovered: int
    discovery_percentage: float
    common_discovered: int
    uncommon_discovered: int
    rare_discovered: int
    legendary_discovered: int
    favorites_count: int
    recent_discoveries: List[int]


class UpdateNotesRequest(BaseModel):
    notes: str = Field(..., description="Personal notes about the discovery")


class QRCodeInfo(BaseModel):
    id: int
    code_token: str
    species_id: int
    location_id: Optional[int] = None
    active: bool
    common_name: Optional[str] = None
    scientific_name: Optional[str] = None
    location_name: Optional[str] = None
    created_at: Optional[datetime] = None


class QRCodeDetail(BaseModel):
    qr_code_id: int
    code_token: str
    species_id: int
    location_id: Optional[int] = None
    active: bool
    is_public: bool = True
    created_at: datetime
    common_name: Optional[str] = None
    scientific_name: str
    location_name: Optional[str] = None
    scan_count: int = 0
    deep_link_url: Optional[str] = None


class QRCodeCreateRequest(BaseModel):
    species_id: int = Field(..., description="The plant species to create a QR code for")
    location_id: Optional[int] = Field(None, description="Optional location ID for the QR code")


class QRCodeBulkCreateRequest(BaseModel):
    species_ids: List[int] = Field(..., description="List of plant species IDs to create QR codes for")
    location_ids: Optional[List[Optional[int]]] = Field(None, description="Optional list of location IDs matching species IDs")


class QRCodeBulkCreateResponse(BaseModel):
    created_count: int
    qr_codes: List[QRCodeInfo]


class ScanQRRequest(BaseModel):
    qr_token: UUID = Field(..., description="The QR code token to scan")

File: app/types/scrapbook.ts
export interface CollectibleCatalogEntry {
  catalog_id: number;
  catalog_number: number;
  plant_species_id: number;
  species_name: string;
  scientific_name: string;
  image_url?: string;
  rarity_tier: 'common' | 'uncommon' | 'rare' | 'legendary';
  featured_order?: number;
  is_discovered: boolean;
  discovered_at?: string;
  discovery_id?: number;
  user_notes?: string;
  is_favorite: boolean;
}

export interface DiscoveryDetails {
  discovery_id: number;
  catalog_number: number;
  plant_species_id: number;
  species_name: string;
  scientific_name: string;
  family?: string;
  rarity_tier: 'common' | 'uncommon' | 'rare' | 'legendary';
  discovered_at: string;
  user_notes?: string;
  is_favorite: boolean;
  plant_article: {
    description?: string;
    care_notes?: string;
    native_habitat?: string;
    conservation_status?: string;
  };
}

export interface DiscoveryResponse {
  success: boolean;
  message: string;
  discovery_id?: number;
  catalog_entry_id?: number;
  catalog_number?: number;
  species_name?: string;
  plant_species_id?: number;
  already_discovered: boolean;
}

export interface CollectionStats {
  total_collectibles: number;
  total_discovered: number;
  discovery_percentage: number;
  common_discovered: number;
  uncommon_discovered: number;
  rare_discovered: number;
  legendary_discovered: number;
  favorites_count: number;
  recent_discoveries: number[];
}

export interface QRCodeInfo {
  id: number;
  code_token: string;
  species_id: number;
  active: boolean;
  common_name?: string;
  scientific_name?: string;
  created_at?: string;
}

